Update reference handling in KLEE `ref<>`

MartinNowack opened the pull request:
* Fixed incorrect handling of copy assignment operator
* Simplify implementing `ref<>`-based objects by moving the implementation into own class
  - simplifies many class structures
* Update `UpdateNode`, `ObjectState` and `MemoryObject` to use ref as well avoids multiple implementations of reference counting
* added move constructor and move assignment operators

kren1 commented on include/klee/util/Ref.h:
```diff
@@ -139,6 +147,19 @@ class ref {
+  // Move assignment operator
+  template <class U> ref<T> &operator=(ref<U> &&r) {
+    // swap
+    auto *tmp = cast_or_null<T>(r.ptr);
```
Why use both `swap` and not swap?

MartinNowack replied:
Here I need to typecast between different types, therefore `std::swap` is less straight forward to use.

kren1 commented on include/klee/Expr.h:
```diff
@@ -450,33 +450,33 @@ class NotOptimizedExpr : public NonConstantExpr {
 public:
-  UpdateNode(const UpdateNode *_next, 
-             const ref<Expr> &_index, 
+  UpdateNode(const ref<UpdateNode> &_next, const ref<Expr> &_index,
```
I don't see the reason for this random line breaks, each parameter on new line makes more sense to me.

kren1 commented on include/klee/util/Assignment.h:
```diff
@@ -75,7 +75,7 @@ namespace klee {
     } else {
       if (allowFreeValues) {
-        return ReadExpr::create(UpdateList(array, 0), 
+        return ReadExpr::create(UpdateList(array, ref<UpdateNode>(nullptr)),
```
Can't this be just `nullptr`?

kren1 commented on lib/Core/Memory.cpp:
```diff
@@ -187,10 +187,10 @@ const UpdateList &ObjectState::getUpdates() const {
     std::vector< std::pair< ref<Expr>, ref<Expr> > > Writes(NumWrites);
-    const UpdateNode *un = updates.head;
-    for (unsigned i = NumWrites; i != 0; un = un->next) {
+    const auto *un = updates.head.get();
```
I think `UpdateNode` is nicer here, but don't want to reheat the `auto` discussions.

kren1 commented on lib/Expr/Updates.cpp:
```diff
@@ -15,37 +15,21 @@ using namespace klee;
-    index(_index),
-    value(_value) {
+UpdateNode::UpdateNode(const ref<UpdateNode> &_next, const ref<Expr> &_index,
+                       const ref<Expr> &_value)
```
Again, much uglier format.

kren1 commented on lib/Solver/STPBuilder.cpp:
```diff
@@ -480,12 +480,11 @@ ::VCExpr STPBuilder::getArrayForUpdate(const Array *root,
-	
-	_arr_hash.hashUpdateNodeExpr(un, un_expr);
+        un_expr =
+            vc_writeExpr(vc, getArrayForUpdate(root, un->next.get()),
```
Much less clear format.

kren1 commented on lib/Solver/STPBuilder.cpp:
```diff
@@ -560,9 +559,9 @@ ExprHandle STPBuilder::constructActual(ref<Expr> e, int *width_out) {
-                       getArrayForUpdate(re->updates.root, re->updates.head),
-                       construct(re->index, 0));
+    return vc_readExpr(
+        vc, getArrayForUpdate(re->updates.root, re->updates.head.get()),
```
Less clear format.

kren1 commented on lib/Core/Memory.h:
```diff
@@ -142,6 +143,22 @@ class MemoryObject {
     }
   }
+
+  int compare(const MemoryObject &b) const {
```
This seems an outlier in this patch.

MartinNowack commented on include/klee/util/Ref.h:
```diff
@@ -7,6 +7,25 @@
+ * struct MyStruct{
+ *   ...
+ *   /// @brief Required by klee::ref-managed objects
+ *   struct ReferenceCounter __refCount;
```
A clear no to default mutability ;)  It's a matter of handling lifetime expactation. The structures should be mutable if you really need them to be, i.e. you need to take partial ownership although the object is marked as `const`. For most of the code, this should not happen. If you need to do this, code should be refactored. Most of the time, a simple reference is better and the more performant solution.

jbuening replied:
Ok, after thinking about this a bit more, I am fine with having to specify whether `__refCount` is mutable or not to make this more explicit in the class that is referenced.

MartinNowack commented:
@kren1, @jbuening  Thanks a lot for your time!! Highly appreciated. I addressed most of the comments. The `operator bool` I would keep for the next time. I repushed the changes. Thanks again.

jbuening commented on include/klee/Expr.h:
```diff
@@ -449,33 +450,33 @@ class NotOptimizedExpr : public NonConstantExpr {
-  
+
+  /// @brief Required by klee::ref-managed objects
+  mutable struct ReferenceCounter __refCount;
```
As you changed `ReferenceCounter` to be a `class`, I would suggest you change it here and in the other places (5 in total).

MartinNowack replied:
@jbuening Thanks for spotting this! Fixed.

jbuening commented:
@MartinNowack thanks for considering our comments. I'm looking forward to see this merged!

MartinNowack commented:
@ccadar This is ready now. Thanks again everybody for the valuable input!

ccadar commented:
@MartinNowack That's a great patch!  Sorry for overlooking it.  Just two comments:
1) Have you checked if this has any significant impact on memory consumption?
2) I really don't like having `auto` everywhere.  I note for instance that in some files that use `UpdateNode`, the name `UpdateNode` doesn't appear at all after your changes!!  I find that this makes the code harder to read, as e.g., now I need to check the type of the container from which the element is extracted to find out the type of the element.

251 commented:
> now I need to check the type of the container from which the element is extracted to find out the type of the element

Sounds like an IDE issue. :grin:

251 commented on include/klee/Expr/Expr.h:
```diff
@@ -543,24 +544,24 @@ class UpdateList {
-  unsigned getSize() const { return (head ? head->getSize() : 0); }
-  
+  unsigned getSize() const {
+    return (head.get() != nullptr ? head->getSize() : 0);
```
Wouldn't it be easier to define `explicit operator bool() const noexcept;` for `ret`?

251 commented on include/klee/Expr/Expr.h:
```diff
@@ -174,7 +174,8 @@ class Expr {
 
-  unsigned refCount;
+  /// @brief Required by klee::ref-managed objects
+  class ReferenceCounter __refCount;
```
Identifiers starting with double underscores are reserved "for implementation" according to C++ standard...

251 commented on include/klee/util/Ref.h:
```diff
@@ -17,32 +36,71 @@ using llvm::cast_or_null;
+
+public:
+  ReferenceCounter() = default;
+  ~ReferenceCounter() = default;
```
Why all the destructors? E.g. this class is trivially destructible.

251 commented on include/klee/util/Ref.h:
```diff
@@ -17,32 +36,71 @@ using llvm::cast_or_null;
 public:
   // default constructor: create a NULL reference
-  ref() : ptr(0) { }
+  ref() : ptr(nullptr) {}
```
Why not via initializer?
```c++
  T *ptr {nullptr};
public:
  ref() = default;
```

ccadar commented:
Hi @MartinNowack.  Let's try to merge this!
I think there are only a couple of things left:
1) Do you want to rename `__refCount` not to use underscores?
2) Can you replace some of the `auto` with the actual types when you think it can help in terms of inline code documentation?
3) Any significant memory impact resulting from replacing an `unsigned` by a `ReferenceCounter` object?

251 commented:
> Any significant memory impact resulting from replacing an unsigned by a ReferenceCounter object?

It has only one member (`unsigned refCount` ) and hence the same size as an `unsigned`.

kren1 commented on include/klee/Expr/Expr.h:
```diff
@@ -449,33 +450,33 @@ class NotOptimizedExpr : public NonConstantExpr {
-  
+
+  /// @brief Required by klee::ref-managed objects
+  mutable class ReferenceCounter _refCount;
```
This is true in the old version too, but why is this mutable and the Expr one isn't? I assume we want ref manged const updateLists, but not Exprs for some reason? (Thinking about this for #1170 )

MartinNowack replied:
You just want to refcount a `const` expression tree. Most often, you don't want to change the expressions at all but you want to share it. Therefore, expr const and only counting itself is non-const.

MartinNowack commented:
@ccadar I think this one is ready as well. I replaced a couple of `auto` with the type specific entries. There is just one underscore now.
Regarding the memory usage, it should be the same. The previous counter is moved into a separate class which becomes a member - no additional allocations.

ccadar commented:
Thanks, @MartinNowack , my bad for not merging this earlier; can you resolve this conflict so that I can do so now?

MartinNowack commented:
@ccadar Ready to be reviewed. I added additional unit tests and fixes from test-comp.

ccadar commented:
Hi @MartinNowack , thanks for the excellent patch!  Really great changes!

I still don't like all the extra `auto` and some of the new formatting, but I'll let it go this time ;)
